Add SOSE potential density task

xylar opened the pull request:
This merge also makes various changes to the script for preprocessing SOSE data:
* switch `depth` to `z` in SOSE output, since it is positive up.  This will mean having to replace SOSE data on all machines once the update is in place. Once we do that, old versions of MPAS-Analysis will break with the new SOSE data. It seems like the change should be made sooner rather than later to avoid this problem once SOSE data is widely used.
* To facilitate computing bottom potential density, `zBot` was added as a coordinate.  A trick was needed to make sure `zBot` gets remapped, since coordinates are not normally remapped.
* There were issues with velocity magnitude on the original lat/lon grid (the 2 components aren't on the same grid) so velocity magnitude is now only available on the Antarctic grid.

xylar commented:
This merge is a branch off of #352 so it should not be merged until that branch has been merged.  Given the changes to the SOSE data, this branch should also only be merged after those changes are in place on the supported machines.

xylar commented:
## Testing and Example Output

RRS30to10wLI:
http://portal.nersc.gov/project/m2833/mpas_analysis_output/sose_20180209.GMPAS-IAF.T62_oRRS30to10v3wLI.cori-knl.afterSalinityFix/ocean/index.html#potDenSose

xylar commented:
@vanroekel and @milenaveneziani, I would appreciate your feedback on the color map and the upper and lower limits of each plot.  Keep in mind that we want values that are useful for all depths so the large range and larger biases at the surface somehow have to be reconciled with the smaller range and biases at depth.

milenaveneziani commented:
Hi @xylar, I have to admit that colormap (`dense`, I suppose?) is not my favorite in `cmocean`: seems too saturated to me and too much of only one color. If you want to stick with `cmocean`, I would try `matter`. Or otherwise my favorite standard colormaps are `magma` or, perhaps better for density, `spectral_r`.

milenaveneziani commented:
Let me experiment a bit myself with the colormaps I mentioned.

milenaveneziani commented:
I am running with this branch, but with `import cmocean` commented out, since I first need to fix my own conda environment before being able to run MPAS-Analysis with it (something to do with netcdf and hdf5 version incompatibility, but that's a different story).
Anyways, after I remove the `import cmocean` bit, I get this error, which reminds me of the python2/3 issue we had before:
```
Traceback (most recent call last):
  File "./run_mpas_analysis", line 565, in <module>
    analyses = build_analysis_list(config, refConfig)
  File "./run_mpas_analysis", line 87, in build_analysis_list
    config=config,  componentName='ocean')
  File "/global/u2/m/milena/MPAS-git-repositories/MPAS-Analysis/mpas_analysis/shared/climatology/ref_year_mpas_climatology_task.py", line 66, in __init__
    config.write(config_string)
  File "/global/project/projectdirs/acme/software/anaconda_envs/edison_e3sm_unified_1.1.3_py2_nox/lib/python2.7/ConfigParser.py", line 407, in write
    fp.write("[%s]\n" % section)
TypeError: unicode argument expected, got 'str'
```

xylar commented:
Ah, right.  This is the *exact* same error cropping up again because this branch (and #355) have not been rebased onto develop after the bug fix.  I'll do that now...

xylar commented:
Here are the surface and bottom PD with the suggested `matter` colormap:
![image](https://user-images.githubusercontent.com/4179064/39955943-ebb8bf02-55d7-11e8-9a49-39057633249e.png)
![image](https://user-images.githubusercontent.com/4179064/39955946-f7fbcdfe-55d7-11e8-88d5-42a889293d06.png)

milenaveneziani commented:
Hi @xylar, I personally like this better than `dense`, I do see more contrast. 
I wanted to try `Spectral_r`, which is what I normally use for density, but I see that the SOSE_pot_den data is not available in the observations folder on edison.

Anyways, yes, we can wait to look at the transects before we finalize this.

xylar commented:
@milenaveneziani, because of the changes I made to the observations (see bullet 1 at the top), there are a new set of observations that will replace the old ones as soon as this gets moved.  They are in:
```
observations/Ocean/SOSE_new
```
on NERSC so just set:
```
[oceanObservations]
...
soseSubdirectory = SOSE_new
```
and it should work.

xylar commented:
Here are runs with `Spectral_R`. 
 I'm not convinced.  I think it's a lot more intuitive to have a colormap that is monotonic in lightness like the `dense` and `matter` maps are.

http://portal.nersc.gov/project/m2833/mpas_analysis_output/sose_20180209.GMPAS-IAF.T62_oEC60to30v3wLI.cori-knl.afterSalinityFix/ocean/index.html#potDenSose

http://portal.nersc.gov/project/m2833/mpas_analysis_output/sose_20180209.GMPAS-IAF.T62_oRRS30to10v3wLI.cori-knl.afterSalinityFix/ocean/index.html#potDenSose

I think I will make `matter` the default at least for now.

milenaveneziani commented:
I do like Spectral_r at the surface. I think we have a more general problem, that the colorbar limits for the surface potential density are never going to be acceptable for the subsurface.
Anyways, we can merge this as is after the new e3sm-unified conda is out, and then change it if we want to.

xylar commented:
@milenaveneziani, in case you want to do some testing yourself, there is an environment you can access with these commands on Edison:
```bash
. /global/project/projectdirs/acme/software/anaconda_envs/${NERSC_HOST}/base/etc/profile.d/conda.sh
conda activate test_e3sm_unified_1.2.0_py2.7_nox
```

milenaveneziani commented:
Since the e3sm-unified environment has been updated, we can merge this one together with #369, right?

milenaveneziani commented:
sounds great.

xylar commented:
Note to self: Don't forget to:
* update the color map to the new default (`Spectral_r`)
* add documentation for the new task following merge of #397 
* Fix `depth` --> `z` in SOSE transects following merge of #327

xylar commented:
I have update the SOSE observations on all supported machines to be the new fields with "z" instead of "depth".  I kept the old files in `SOSE_old` in case people have trouble and need the old ones.  But I'll delete them after the v1.0 release.

xylar commented:
I re-tested everything on my laptop and on Edison.   I updated the documentation to include the new task (and fixed a few missing pieces elsewhere).  I'm going to go ahead a merge as soon as those tests and CI have finished.
